spec/GraphQLOverHTTP.md

```diff
@@ -215,3 +215,31 @@ as long as the server is still able to produce a well-formed response.
 
 In case of errors that completely prevent the successful execution of the request,
 the server SHOULD respond with the appropriate status code depending on the concrete error condition.
+
+The following examples provide guidance on how to deal with specific error cases:
+
+### Unparseable or invalid request body
+
+For example: `NONSENSE`, `{"qeury": "{__typena`
+
+Completely prevents execution of the GraphQL operation and SHOULD result in status code `4ß0` (Bad Request).
+
+### Document validation
+
+Includes validation steps that run before execution of the GraphQL operation:
+- [GraphQL specification validation](http://spec.graphql.org/June2018/#sec-Validation)
+- custom validation, for example: depth limit, complexity limit
+
+The server MAY choose to:
+- deny execution, the status code SHOULD be `400` (Bad Request)
+- attempt execution, the status code will depend on the validity of the result
```

Comment:
Why are we allowing the server to attempt execution when validation fails? Surely that could lead to a whole host of issues (for just one example, infinite recursion from recursive fragments might be permitted if just that validation rule failed). Is it specific to custom validation rules?

I'm not saying we shouldn't (though that is the direction I'm leaning); but I'd like to understand the reasons why we think we should (and I'd like that added as a non-normative note).

Reply:
@benjie - Let me know if I'm missing your point.  I certainly support a non-normative note, but I don't think it's the job of the `GraphQL over HTTP` spec to allow or not allow execution on an invalid query.  The `GraphQL` spec itself says:

> An invalid request is still technically executable, and will always produce a stable result as defined by the algorithms in the Execution section, however that result may be ambiguous, surprising, or unexpected relative to a request containing validation errors, so execution should only occur for valid requests.

I read that as "you *should* only execute valid queries, but you *may* execute invalid queries".  I might be missing somewhere else in the spec that says differently?  If I'm correct and if every legal GraphQL server may be used over `GraphQL over HTTP`, then we should be controlling how they reply over HTTP whether or not they choose to execute an invalid query, exactly as long as the spec allows it.

While I generally always support vaIidation before execution, in the "real world" I think it's reasonable to not validate queries that you know to be valid - for example if they are generated by your code or already validated by other middleware - and then what if there's a bug and it wasn't valid after all?  I think that our spec should be resilient enough to explain simple rules that are still relevant like "any non-null data must have a 2xx status code" etc.

Even if you agree with that, I would definitely support rewriting the text to be clearer what we're saying and not imply that we recommend executing invalid queries.

Reply:
I think the first bullet should be SHOULD rather than MAY. At the moment it sounds like they’re both equivalent.

Though the GraphQL spec does indeed talk about what is _technically_ possible (this is necessary to help library authors structure their GraphQL libraries), I do think the GraphQL over HTTP Spec should be recommending what users should do when exposing a GraphQL schema over HTTP, and validation is essential to that (whether it’s done when the request is received, or beforehand e.g. during a build phase with persisted operations, or cached from the last time they saw that same document).

Reply:
I think I'm happy with SHOULD there.  I totally agree that they're not equal clauses.

Brainstorming - Might it be better instead to rewrite to avoid mentioning it?  I think our spec might be stronger if it only mentions internals for the GraphQL execution engine in non-normative notes to help implementers, but for the requirements sticks to things like "a non-null `data` MUST have a 2xx, and SHOULD have a 200."  (Then leaving the internals of what execution engine paths yield that non-null `data` in the first place to the GraphQL spec as well as the non-normative notes/section.)

For your other point, I agree that validation is essential - often it's in one of the ways you mention but sometimes at our customers it's done at a different point in the network topology and the current server has a "guarantee" that the incoming transactions are valid (internal generated queries, or already validated in the DMZ, etc.), and obviously sometimes those "guarantees" fail.  That's not universal, and not limited to GraphQL, but I think it's a real tradeoff out there.

Reply:
I just re-checked the PR and I see two things I'd forgotten (sorry I didn't check earlier):
- 1. this is already inside an 'example' and thus I assume this whole part of the text is non-normatime.  As such, I'm happy for it to say SHOULD, but I think it could be stronger, and say something like "the server would probably return a 4xx" or something like that and not even mention the other option.
- 2. it explicitly lists multiple types of "bad request", one being a strict spec failure to validate and the other being "custom validation, for example: depth limit, complexity limit" - for these kinds of 'extra' checks I think that both "log a warning and continue" and "error immediately" are acceptable options.  I don't know who originally wrote that text but maybe that was the motivation.

Reply:
Thank you for the feedback, I think it is generally a good recommendation to abort. I came up with a use case for continuing regardless: persisted queries that were previously known to be valid. I adapted this section accordingly: https://github.com/graphql/graphql-over-http/pull/125/commits/5520f071f64c9fad6145bc3f49b1e16504f873b0